### PR TITLE
Add TestInjectionPolicy inference unittest class for testing custom injection policies

### PR DESCRIPTION
**Pull request by lekurile**
This PR adds a `TestInjectionPolicy` inference unittest class for testing custom injection policies.

This test differs from the existing tests in that the `injection_policy` dictionary is explicitly specified when calling the DeepSpeed `init_inference` API.

The `google/t5-v1_1-small` text2text-generation model and the `roberta-large` fill-mask model are added as tests with the injection policy explicitly specified.

This is done to expand our unittest coverage to test the path where the `replace_wo_policy` function is invoked (see GH-2387).

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -2,10 +2,13 @@
 
 import sys
 import pytest
+import os
 from os.path import abspath, dirname, join
 import torch
 import warnings
 
+os.environ['PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION'] = 'python'
```
Comment (mrwyattii) [inline tests/conftest.py:10]: Can we add a comment above this line indicating for which test this is needed?

File: tests/unit/inference/test_inference.py [hunk 2]
```diff
@@ -252,14 +266,28 @@ def test(
...
+                pipe.model,
+                mp_size=1,
+                dtype=dtype,
+                injection_policy={
+                    T5Block: ('SelfAttention.o',
+                              'EncDecAttention.o',
+                              'DenseReluDense.wo')
+                },
+                enable_cuda_graph=enable_cuda_graph,
+            )
```
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:280]: Adding additional if statements to the main test function could become messy. I think this is ok for now, but we should think about how we can have a single `init_inference` call.

**Review by mrwyattii — APPROVED** [review]
Add a comment to the one line and then good to merge!

File: tests/conftest.py (hunk 1 above)
Comment (lekurile) [reply tests/conftest.py:11]: Added a comment describing that this is needed for the T5 unittest.

File: tests/unit/inference/test_inference.py (hunk 2 above)
Comment (lekurile) [reply tests/unit/inference/test_inference.py]: Thanks for the suggestion. After our discussion, we decided to refactor the testing to include a `TestInjectionPolicy` class, specifically designed to test explicit injection policies.

**Review by awan-10 — APPROVED** [review]
Great!

File: tests/unit/inference/test_inference.py [hunk 3]
```diff
@@ -323,6 +334,66 @@ def test(
...
+        (("roberta-large",
+          "fill-mask"),
+         {
+             RobertaLayer: ('output.dense')
+         }),
+    ],
+    ids=["t5",
+         "roberta"],
+)
+@pytest.mark.parametrize("world_size", [1, 2])
```
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:357]: I'm surprised this works for our `DistributedTest` class, as it was certainly not designed to work this way. The `world_size` parameter should be a property of the class. See https://github.com/microsoft/DeepSpeed/blob/799120e7e44cefde7db8005295d40c2fa1eaf28d/tests/unit/common.py#L82

File: tests/unit/inference/test_inference.py [hunk 4]
```diff
@@ -323,6 +334,66 @@ def test(
...
+@pytest.mark.parametrize("dtype", [torch.float], ids=["fp32"])
+@pytest.mark.parametrize("enable_cuda_graph", [False], ids=["noCG"])
```
Comment (mrwyattii) [inline tests/unit/inference/test_inference.py:359]: These are not required as the `dtype` and `enable_cuda_graph` in the test method definition have default values.

**Review by mrwyattii — CHANGES_REQUESTED** [review]

File: tests/unit/inference/test_inference.py (hunk 3 above)
Comment (lekurile) [reply tests/unit/inference/test_inference.py]: Updated to no longer parameterize the world_size.

File: tests/unit/inference/test_inference.py (hunk 4 above)
Comment (lekurile) [reply tests/unit/inference/test_inference.py:358]: Removed the defaults from the test method and kept parametrize to override the fixtures.

**Review by mrwyattii — APPROVED** [review]
